perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java

```diff
@@ -3029,6 +3037,15 @@ public void deleteAttribute(PerunSession sess, AttributeDefinition attributeDefi
 		return attributes;
 	}
 
+	@Override
+	public List<Attribute> getRequiredAttributes(PerunSession sess, List<Service> services, Facility facility, Resource resource, User user, Member member) throws InternalErrorException, MemberResourceMismatchException {
+		List<Attribute> attributes = new ArrayList<>();
+		for (Service service : services) {
+			getRequiredAttributes(sess, service, facility, resource, user, member);
```

Comment:
I believe you want to add result into the ``attributes`` list. Also, what if required attributes of passed services overlap? We probably should gather them into the set before returning as list.

Reply:
Yeah, definitely I wanted. I have fixed it and also added a test for this method.

Reply:
When I think about it, our ```equals()``` include attribute value. So if value of required attribute changes during the consequent calls, we still might get duplicity, which will probably fails some checks. But optimization of impl layer to load all data from DB at once is not a part of this change.